docs/docs/debugging-incremental-builds.md

```diff
@@ -0,0 +1,87 @@
+
+- which page template the page is using
+- results of page query
+- results of static queries used by a page template
+- frontend source code (shared and also browser `gatsby-browser.js` / SSR `gatsby-ssr.js` specifically)
+
+When those inputs change since the last build, the HTML files are marked to be regenerated. If they don't change, Gatsby can reuse HTML files generated in previous build.
+
+## Debugging
+
+To be able to see the changed pages inside console, run `gatsby build` with some additional flags:
+
+```shell
+gatsby build --verbose --log-pages
+```
+
+At the end of the build you will see the changed pages:
+
+```shell
+info Built pages:
+Updated page: /foo/bar/
+```
+
+If you also want to see what and how files actually did change between builds you can look at the `diff` between two builds. Inside your project directory run these commands:
+
+```shell
+gatsby clean
+gatsby build --no-uglify
+cp -r public public-first-build
+```
+
+Now make your changes to the files and continue with these commands:
+
+```shell
+gatsby build --no-uglify --verbose --log-pages
+diff -u -r public-first-build public > build-diff.txt
+```
+
+Open the `build-diff.txt` file in the root of your site to see the difference between both builds.
+
+## Tips
+
+### Expected behaviors
+
+There are certain changes that you can make to your project that trigger more changes than anticipated or even a full rebuild. For the following cases this is expected behavior:
+
+- Changing a React component / page. This results in a changed webpack compilation where some assets generated by webpack will have changed hashes in filenames and thus Gatsby needs to regenerate pages to update `<link>` and `<script>` references.
+
+### Avoid direct filesystem calls
+
+As we mentioned, Gatsby tracks "inputs" used to generate HTML files. However, the `gatsby-ssr` file allows some arbitrary code execution like using the [`fs` module](https://nodejs.org/api/fs.html). For example:
+
+```js:title=gatsby-ssr.js
+const fs = require(`fs`)
+const someUntrackedInput = fs.readFileSync(`some-path.txt`)
+
+// Rest of gatsby-ssr.js file
+```
+
+While Gatsby could also track files that are read, the custom code that does those reads might have some special logic that Gatsby is not aware of. In the above example the filename could be generated and changed between builds, or the file read itself could change -- so Gatsby sees this as "arbitrary" file reading. If Gatsby discovers that `fs` modules are used, it will disable "Incremental Builds"-mode to stay on the safe side (there will be warnings mentioning "unsafe builtin method").
+
+If your `gatsby-ssr` (either site itself or plugin) make use of `fs` reads, head over to [migrating from v2 to v3 guide](/docs/reference/release-notes/migrating-from-v2-to-v3/#using-fs-in-ssr) and check how to migrate.
+
+### Avoid calling date stuff
```

Comment:
stuff ;) 

I would we could try to be more specific and pinpoint to "current" date. Dates are fine - especiall for cases like dates of blog posts and what not - I don't want to give wrong impression here that all dates are bad for inc builds.
```suggestion
### Avoid relying on current date and/or time
```